### PR TITLE
[calendar] Fixed `createEventAsync` crashing when `alarms` were set or `endTimeZone` was null

### LINKED ISSUE #9184: expo-calendar can't create events with notification alarms on Android
### Summary of Issue
I can't create an event with alarms on Android.  Depending on the attributes to the createEventAsync event object, I get a crash or error.  This happens with both all day events and non-all-day.  It happens if the relativeOffset is in the past or future.  It used to work in expo-calendar 8.1.0

### Environment 
Android, with expo-calendar 8.2.1

```
  Expo CLI 3.22.1 environment info:
    System:
      OS: Windows 10 10.0.18362
    Binaries:
      Node: 10.15.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.22.0 - C:\Users\joshr\AppData\Roaming\npm\yarn.CMD
      npm: 6.4.1 - C:\Program Files\nodejs\npm.CMD
    IDEs:
      Android Studio: Version  4.0.0.0 AI-193.6911.18.40.6514223
    npmPackages:
      expo: ^38.0.0 => 38.0.8
      react: 16.11.0 => 16.11.0
      react-dom: 16.11.0 => 16.11.0
      react-native: https://github.com/expo/react-native/archive/sdk-38.0.2.tar.gz => 0.62.2
      react-native-web: ~0.11.7 => 0.11.7
```

### Reproducible Demo

This causes a crash. 
```
    const start = new Date(new Date().setUTCHours(0, 0, 0, 0));
    const end = new Date(new Date(start.getTime() + 24 * 60 * 60000).setUTCHours(0, 0, 0, 0));
    const newCalendarEvent = {
      alarms: [{ relativeOffset: 9 * 60, method: Calendar.AlarmMethod.ALERT }],
      notes: "This is an all-day test event",
      title: "Test event",
      startDate: start,
      endDate: end,
      allDay: true,
      timeZone: "UTC",
      endTimeZone: "UTC",
    };
    const eventId = await Calendar.createEventAsync(calendarId, newCalendarEvent);
```
In a previous release of expo 37 and expo-calendar 8.1.0 I had code like this, but with endTimeZone set to null and it worked.  In release 8.2.1, endTimeZone: null causes a "TypeError: Cannot convert undefined or null to object".

I probably had null because that's how they appear when created by the google calendar app.  I recall having problems (maybe the crash existed, I don't remember) and I got things to work after making my object look similar to events created by the google calendar app.  

The [stringifyDateValues ](https://github.com/expo/expo/blob/master/packages/expo-calendar/src/Calendar.ts#L702) function was recently changed.  Null is an object.  This function should be changed to return null if obj is null.

 
### Steps to Reproduce
To get the crash, attempt to create an all day calendar entry with an alarm with the code above.
To get the other error, try the same thing with endTimeZone: null

### Expected Behavior vs Actual Behavior
Expected: Calendar entry created without crash whether endTimeZone is specified, not specified, or specified as null.  

Note that it's good to support null because that's how the objects appear that are created by the default calendar app.  If an expo app wants to query and clone an all day event created with the default calendar app it would have endTimeZone: null.

Actual: Crash or error.

### PR DESCRIPTION
**Pull request by barthap**
# Why
 - Fixes #9184 
 - Fixes invalid TypeScript type for `RecurrenceRule.endDate` - it was inconsistent with docs.

# How
Modified `stringifyDateValues()` in `Calendar.ts`:
 - Added support for situation, when object value is null (now nullable `Event.endTimeZone` is properly handled).
Because `typeof null = "object"`, _null_ was passed to `Object.keys()` and it threw an exception.
- Added support for array values - now `Event.alarms` is properly handled.

# Test Plan
Tested by running and debugging code example from issue.

### CODE REVIEW

**Review by bbarthec — CHANGES_REQUESTED** [review]
Good job 👍 But let's approach it a little different 😉

File: packages/expo-calendar/src/Calendar.ts [hunk 1]
```diff
@@ -699,10 +699,17 @@ function stringifyIfDate(date: any): any {
...
+  return arr.map(obj => stringifyDateValues(obj));
+}
+
 function stringifyDateValues(obj: object): object {
   return Object.keys(obj).reduce((acc, key) => {
     const value = obj[key];
-    if (typeof value === 'object' && !(value instanceof Date)) {
+    if (value != null && typeof value === 'object' && !(value instanceof Date)) {
+      if (Array.isArray(value)) {
+        return { ...acc, [key]: stringifyArrayValues(value) };
```
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-calendar/src/Calendar.ts:711]: Yep, you're right - I thought of inlining this code but I was wondering about readibility 😉

File: packages/expo-calendar/CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,7 @@
 ### 🐛 Bug fixes
 
 - Fix `Calendar.getEventsAsync` crashing when `recurrenceRules` are incorrect. ([#8760](https://github.com/expo/expo/pull/8760) by [@lukmccall](https://github.com/lukmccall))
+- Fixed `stringifyDate` helper functions. ([#9269](https://github.com/expo/expo/pull/9269) by [@barthap](https://github.com/barthap))
```
Comment (tsapeta) [inline packages/expo-calendar/CHANGELOG.md:14]: This changelog entry would be good and clear for us internally, but not really for our users as the `stringifyDate` is not being exported by the module and it's not a part of our API. Could you rephrase it to be more about user-facing changes (an issue that this PR will close might be helpful)?

**Review by tsapeta — APPROVED** [review]
Thanks, looks good besides the changelog entry 👍 
Keep up the good work! 😉

File: packages/expo-calendar/CHANGELOG.md (hunk 2 above)
Comment (barthap) [reply packages/expo-calendar/CHANGELOG.md]: I think this should be better now
https://github.com/expo/expo/pull/9269/files#diff-75d1dd6f04ae24a7dfb0e11eb467376a

**Comment by r20** [comment]
Sorry, I'm new and not familiar with this.  But, will there be a way to get this fix before expo SDK 39?  (If I have to wait until September, I'll downgrade back to 37 where I can avoid the crash.)

Also, was the fix tested?  The fix description says the code was changed to handle alarms in an array and endTimeZone as null.  The crash occurs with endTimeZone not null and alarms not in an array.  I suspect there is yet another bug besides the null and array issues.

**Comment by barthap** [comment]
@r20 The crash occured when `alarms` were set OR when any nullable event parameter (including `endTimeZone`) was set to null. According to [`expo-calendar` docs](https://docs.expo.io/versions/v38.0.0/sdk/calendar/#calendarcreateeventasynccalendarid-details), `alarms` should always be array. It may be unspecified, empty, or contain one or more `Alarm` objects.

> But, will there be a way to get this fix before expo SDK 39?

Unfortunately, I don't know that.

**Comment by tsapeta** [comment]
@r20 I've just published `expo-calendar@8.2.2` that can be used in managed workflow with SDK38. Please let us know if it's been fixed for you 🙂 

I'm also going to publish version `8.3.0` for bare workflow as there were some other bug fixes in the meantime but unfortunately they wouldn't work with SDK38.

**Comment by r20** [comment]
@tsapeta Yes, it looks like it's fixed with 8.2.2.  I am able to use null and non-null for endTimeZone and set alarms.  No errors, no crashes. Thanks.
